### PR TITLE
implement of the engine and rewrite factory

### PR DESCRIPTION
**Pull request by LHT129**
issue: #203

- allocator shared_ptr hold in Index and Dataset
- the allocator ptr always point to SafeAllocator
- allocator ptr is used inner index
- only Engine/Resource can create Allocator Object

### CODE REVIEW

File: include/vsag/engine.h [hunk 1]
```diff
@@ -17,20 +17,26 @@
...
     void
-    Shutdown();
+    Shutdown();  // like ~Engine(), but will warn whether the resources are still reference outside
 
     tl::expected<std::shared_ptr<Index>, Error>
     CreateIndex(const std::string& name, const std::string& parameters);
 
+    //    tl::expected<DatasetPtr, Error>
+    //    CreateDataset();
+
```
Comment (inabao) [inline include/vsag/engine.h]: delete unnecessary modifications

File: src/algorithm/hgraph.cpp [hunk 2]
```diff
@@ -43,14 +43,14 @@ empty_binaryset() {
...
-      labels_(common_param.allocator_) {
+      label_lookup_(common_param.allocator_.get()),
+      label_op_mutex_(MAX_LABEL_OPERATION_LOCKS, common_param.allocator_.get()),
+      neighbors_mutex_(0, common_param.allocator_.get()),
+      route_graphs_(common_param.allocator_.get()),
+      labels_(common_param.allocator_.get()) {
     this->dim_ = common_param.dim_;
     this->metric_ = common_param.metric_;
-    this->allocator_ = common_param.allocator_;
+    this->allocator_ = common_param.allocator_.get();
```
Comment (jiaweizone) [inline src/algorithm/hgraph.cpp:63]: HGraph can hold the allocator shared_ptr ?

File: src/factory/factory.cpp [hunk 3]
```diff
@@ -13,87 +13,30 @@
...
-    } catch (const std::invalid_argument& e) {
-        LOG_ERROR_AND_RETURNS(
-            ErrorType::INVALID_ARGUMENT, "failed to create index(invalid argument): ", e.what());
-    } catch (const std::exception& e) {
-        LOG_ERROR_AND_RETURNS(
-            ErrorType::UNSUPPORTED_INDEX, "failed to create index(unknown error): ", e.what());
-    }
+    Resource resource(allocator);
+    Engine e(&resource);
+    return e.CreateIndex(origin_name, parameters);
```
Comment (jiaweizone) [inline src/factory/factory.cpp:39]: Here only need to hold `allocator` object ?

File: src/io/memory_io.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ class MemoryIO : public BasicIO<MemoryIO> {
     }
 
     MemoryIO(const JsonType& io_param, const IndexCommonParam& common_param)
-        : allocator_(common_param.allocator_) {
+        : allocator_(common_param.allocator_.get()) {
```
Comment (jiaweizone) [inline src/io/memory_io.h:39]: ditto

File: src/safe_allocator.h [hunk 5]
```diff
@@ -17,18 +17,21 @@
...
 #include "vsag/allocator.h"
 
 namespace vsag {
 
 class SafeAllocator : public Allocator {
 public:
-    explicit SafeAllocator(Allocator* raw_allocator, bool owner = false)
-        : raw_allocator_(raw_allocator), owner_(owner) {
+    static std::shared_ptr<Allocator>
+    FactoryDefaultAllocator() {
```
Comment (jiaweizone) [inline src/safe_allocator.h:28]: FactoryDefaultAllocator -> NewDefaultAllocator ?

File: src/algorithm/hgraph.cpp (hunk 2 above)
Comment (LHT129) [reply src/algorithm/hgraph.cpp:63]: Hgraph Index hold the shared ptr

File: src/factory/factory.cpp (hunk 3 above)
Comment (LHT129) [reply src/factory/factory.cpp:39]: maybe threadpool or other resource in future

File: src/io/memory_io.h (hunk 4 above)
Comment (LHT129) [reply src/io/memory_io.h:39]: the MemoryIO's lifetime is shorter than Index, so only raw pointer is enough

File: src/safe_allocator.h (hunk 5 above)
Comment (LHT129) [reply src/safe_allocator.h:28]: Both are possible, but the former is a verb-object structure and is more suitable as a function name.

File: src/algorithm/hgraph.cpp (hunk 2 above)
Comment (jiaweizone) [reply src/algorithm/hgraph.cpp:63]: It is recommended use shared_ptr to avoid possible memory issues and keep consistent allocator object passing logic

Comment (LHT129) [reply src/algorithm/hgraph.cpp:63]: if lifetime is shorter than Index, only raw pointer is enough, too much shared_ptr's construction will influence the performance

**Review by wxyucs — COMMENTED** [review]
@LHT129 we need to add an example for engine.

File: include/vsag/engine.h (hunk 1 above)
Comment (LHT129) [reply include/vsag/engine.h]: done

File: src/factory/factory.cpp [hunk 6]
```diff
@@ -13,87 +13,30 @@
...
 
-#include "index/diskann.h"
-#include "index/diskann_zparameters.h"
-#include "index/hgraph_index.h"
-#include "index/hgraph_zparameters.h"
-#include "index/hnsw.h"
-#include "index/hnsw_zparameters.h"
-#include "index/index_common_param.h"
-#include "vsag/vsag.h"
+#include "ThreadPool.h"
```
Comment (inabao) [inline src/factory/factory.cpp:27]: why include it

**Review by inabao — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Review by ShawnShawnYou — APPROVED** [review]

File: src/algorithm/hgraph.cpp (hunk 2 above)
Comment (jiaweizone) [reply src/algorithm/hgraph.cpp:63]: Ok, we'll discuss this later

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**